File: mobile/shared/src/commonMain/kotlin/uk/foodclassifier/image/ImageCropScreen.kt
package uk.foodclassifier.image

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeContentPadding
import androidx.compose.foundation.Image
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathFillType
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp

@Composable
fun ImageCropScreen(
    imageBytes: ByteArray,
    initialCrop: NormalizedRect = NormalizedRect.ingredientsGuide,
    onConfirm: (ByteArray) -> Unit,
    onRetake: () -> Unit,
    onCancel: () -> Unit,
) {
    val bitmap = remember(imageBytes) { decodeImageBitmap(imageBytes) }
    if (bitmap == null) {
        onCancel()
        return
    }

    var cropRect by remember(imageBytes, initialCrop) { mutableStateOf(initialCrop) }

    Surface(
        modifier = Modifier.fillMaxSize(),
        color = MaterialTheme.colorScheme.surface,
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .safeContentPadding(),
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 4.dp, vertical = 4.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TextButton(onClick = onCancel) {
                    Text("Cancel")
                }
                TextButton(onClick = onRetake) {
                    Text("Retake")
                }
            }

            CropImageCanvas(
                bitmap = bitmap,
                cropRect = cropRect,
                onCropRectChange = { cropRect = it },
                modifier = Modifier
                    .weight(1f)
                    .fillMaxWidth()
                    .padding(horizontal = 12.dp),
            )

            Button(
                onClick = {
                    val cropped = cropImageBytes(imageBytes, cropRect) ?: imageBytes
                    onConfirm(cropped)
                },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
            ) {
                Text("Scan")
            }
        }
    }
}

@Composable
internal fun CropGuideOverlay(
    guide: NormalizedRect,
    modifier: Modifier = Modifier,
    contentWidth: Float? = null,
    contentHeight: Float? = null,
) {
    Canvas(modifier = modifier) {
        val contentBounds = if (contentWidth != null && contentHeight != null) {
            fitImageBounds(size.width, size.height, contentWidth, contentHeight)
        } else {
            ImageBounds(0f, 0f, size.width, size.height)
        }
        val screenRect = contentBounds.toScreenRect(guide)
        val hole = Rect(
            offset = Offset(screenRect.left, screenRect.top),
            size = Size(
                width = screenRect.right - screenRect.left,
                height = screenRect.bottom - screenRect.top,
            ),
        )
        val overlayPath = Path().apply {
            fillType = PathFillType.EvenOdd
            addRect(Rect(Offset.Zero, size))
            addRect(hole)
        }
        drawPath(overlayPath, color = Color.Black.copy(alpha = 0.55f))
        drawRect(
            color = Color.White,
            topLeft = hole.topLeft,
            size = hole.size,
            style = Stroke(width = 3.dp.toPx()),
        )
    }
}

@Composable
private fun CropImageCanvas(
    bitmap: ImageBitmap,
    cropRect: NormalizedRect,
    onCropRectChange: (NormalizedRect) -> Unit,
    modifier: Modifier = Modifier,
) {
    BoxWithConstraints(
        modifier = modifier.background(Color.Black, androidx.compose.foundation.shape.RoundedCornerShape(12.dp)),
    ) {
        val imageBounds = fitImageBounds(
            containerWidth = constraints.maxWidth.toFloat(),
            containerHeight = constraints.maxHeight.toFloat(),
            imageWidth = bitmap.width.toFloat(),
            imageHeight = bitmap.height.toFloat(),
        )

        Box(modifier = Modifier.fillMaxSize()) {
            Image(
                bitmap = bitmap,
                contentDescription = "Photo to crop",
                modifier = Modifier.fillMaxSize(),
                contentScale = ContentScale.Fit,
            )

            Canvas(
                modifier = Modifier
                    .fillMaxSize()
                    .pointerInput(imageBounds, cropRect) {
                        detectDragGestures { change, dragAmount ->
                            change.consume()
                            if (imageBounds.width <= 0f || imageBounds.height <= 0f) {
                                return@detectDragGestures
                            }
                            onCropRectChange(
                                cropRect.translated(
                                    normalizedDeltaX = dragAmount.x / imageBounds.width,
                                    normalizedDeltaY = dragAmount.y / imageBounds.height,
                                ),
                            )
                        }
                    },
            ) {
                val screenRect = imageBounds.toScreenRect(cropRect)
                val hole = Rect(
                    offset = Offset(screenRect.left, screenRect.top),
                    size = Size(
                        width = screenRect.right - screenRect.left,
                        height = screenRect.bottom - screenRect.top,
                    ),
                )
                val overlayPath = Path().apply {
                    fillType = PathFillType.EvenOdd
                    addRect(Rect(Offset.Zero, size))
                    addRect(hole)
                }
                drawPath(overlayPath, color = Color.Black.copy(alpha = 0.55f))
                drawRect(
                    color = Color.White,
                    topLeft = hole.topLeft,
                    size = hole.size,
                    style = Stroke(width = 3.dp.toPx()),
                )
            }
        }
    }
}

File: flutter/lib/screens/image_crop_screen.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:food_classifier/image/image_crop_utils.dart';
import 'package:food_classifier/image/normalized_rect.dart';
import 'package:image/image.dart' as img;

class ImageCropScreen extends StatefulWidget {
  const ImageCropScreen({
    super.key,
    required this.imageBytes,
    required this.initialCrop,
    required this.onConfirm,
    required this.onRetake,
    required this.onCancel,
  });

  final Uint8List imageBytes;
  final NormalizedRect initialCrop;
  final ValueChanged<Uint8List> onConfirm;
  final VoidCallback onRetake;
  final VoidCallback onCancel;

  @override
  State<ImageCropScreen> createState() => _ImageCropScreenState();
}

class _ImageCropScreenState extends State<ImageCropScreen> {
  late NormalizedRect _cropRect;

  @override
  void initState() {
    super.initState();
    _cropRect = widget.initialCrop;
  }

  @override
  void didUpdateWidget(covariant ImageCropScreen oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.imageBytes != widget.imageBytes ||
        oldWidget.initialCrop != widget.initialCrop) {
      _cropRect = widget.initialCrop;
    }
  }

  void _confirm() {
    final cropped = cropImageBytes(widget.imageBytes, _cropRect) ?? widget.imageBytes;
    widget.onConfirm(cropped);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                TextButton(onPressed: widget.onCancel, child: const Text('Cancel')),
                TextButton(onPressed: widget.onRetake, child: const Text('Retake')),
              ],
            ),
            Expanded(
              child: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 12),
                child: _CropImageCanvas(
                  imageBytes: widget.imageBytes,
                  cropRect: _cropRect,
                  onCropRectChange: (rect) => setState(() => _cropRect = rect),
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(16),
              child: FilledButton(
                onPressed: _confirm,
                child: const Text('Scan'),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _CropImageCanvas extends StatelessWidget {
  const _CropImageCanvas({
    required this.imageBytes,
    required this.cropRect,
    required this.onCropRectChange,
  });

  final Uint8List imageBytes;
  final NormalizedRect cropRect;
  final ValueChanged<NormalizedRect> onCropRectChange;

  Size _imageDisplaySize() {
    final decoded = img.decodeImage(imageBytes);
    if (decoded == null) {
      return Size.zero;
    }
    final baked = img.bakeOrientation(decoded);
    return Size(baked.width.toDouble(), baked.height.toDouble());
  }

  @override
  Widget build(BuildContext context) {
    final imageSize = _imageDisplaySize();

    return LayoutBuilder(
      builder: (context, constraints) {
        final imageBounds = imageSize == Size.zero
            ? ImageBounds(
                left: 0,
                top: 0,
                width: constraints.maxWidth,
                height: constraints.maxHeight,
              )
            : fitImageBounds(
                containerWidth: constraints.maxWidth,
                containerHeight: constraints.maxHeight,
                imageWidth: imageSize.width,
                imageHeight: imageSize.height,
              );

        return ClipRRect(
          borderRadius: BorderRadius.circular(12),
          child: ColoredBox(
            color: Colors.black,
            child: Stack(
              fit: StackFit.expand,
              children: [
                Image.memory(
                  imageBytes,
                  fit: BoxFit.contain,
                ),
                GestureDetector(
                  onPanUpdate: (details) {
                    if (imageBounds.width <= 0 || imageBounds.height <= 0) {
                      return;
                    }
                    onCropRectChange(
                      cropRect.translated(
                        normalizedDeltaX: details.delta.dx / imageBounds.width,
                        normalizedDeltaY: details.delta.dy / imageBounds.height,
                      ),
                    );
                  },
                  child: CustomPaint(
                    painter: _CropOverlayPainter(
                      cropRect: cropRect,
                      imageBounds: imageBounds,
                    ),
                    child: const SizedBox.expand(),
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}

class _CropOverlayPainter extends CustomPainter {
  _CropOverlayPainter({
    required this.cropRect,
    required this.imageBounds,
  });

  final NormalizedRect cropRect;
  final ImageBounds imageBounds;

  @override
  void paint(Canvas canvas, Size size) {
    final screenRect = imageBounds.toScreenRect(cropRect);
    final hole = Rect.fromLTRB(
      screenRect.left,
      screenRect.top,
      screenRect.right,
      screenRect.bottom,
    );

    final overlayPath = Path()
      ..fillType = PathFillType.evenOdd
      ..addRect(Offset.zero & size)
      ..addRect(hole);

    canvas.drawPath(
      overlayPath,
      Paint()..color = Colors.black.withValues(alpha: 0.55),
    );
    canvas.drawRect(
      hole,
      Paint()
        ..color = Colors.white
        ..style = PaintingStyle.stroke
        ..strokeWidth = 3,
    );
  }

  @override
  bool shouldRepaint(covariant _CropOverlayPainter oldDelegate) {
    return oldDelegate.cropRect != cropRect ||
        oldDelegate.imageBounds.left != imageBounds.left ||
        oldDelegate.imageBounds.top != imageBounds.top ||
        oldDelegate.imageBounds.width != imageBounds.width ||
        oldDelegate.imageBounds.height != imageBounds.height;
  }
}
